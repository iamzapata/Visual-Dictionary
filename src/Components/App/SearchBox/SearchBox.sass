.SearchBox
  display: flex
  align-items: center
  justify-content: space-between