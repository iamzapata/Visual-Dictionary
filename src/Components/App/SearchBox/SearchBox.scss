.SearchBox display: flex align-items: center justify-content: space-between
  .SearchBox__Form width: 100% min-height: 75px .SearchBox__Input-Control flex:
  1 .ErrorMessage letter-spacing: 1px text-align: center .Suggestions margin:
  10px 0;
