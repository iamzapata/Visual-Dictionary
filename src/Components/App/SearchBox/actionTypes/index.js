import keyMirror from 'keymirror'

const ActionTypes = keyMirror({
  SEARCH_WORD_REQUEST: null,
  SEARCH_WORD_SUCCESS: null,
  SEARCH_WORD_FAILURE: null
})

export default ActionTypes