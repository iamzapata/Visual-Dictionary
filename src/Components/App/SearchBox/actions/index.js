import ActionTypes from "Components/App/SearchBox/actionTypes"
import request from "utils/request"

export const searchWordRequest = searchQuery => ({
  type: ActionTypes.SEARCH_WORD_REQUEST,
  err: null,
  isLoading: true,
  searchQuery
})

export function searchWordSuccess(response) {
  const { results } = response
  return {
    type: ActionTypes.SEARCH_WORD_SUCCESS,
    err: null,
    isLoading: false,
    results
  }
}

export function searchWordFailure(err) {
  return {
    type: ActionTypes.SEARCH_WORD_FAILURE,
    err,
    isLoading: false
  }
}

export default function searchWord(queryString) {
  return dispatch => {
    dispatch(searchWordRequest(queryString))
    return request(`entries/en/${queryString}`)
      .then(response => {
        dispatch(searchWordSuccess(response))
      })
      .catch(err => {
        dispatch(searchWordFailure(err))
      })
  }
}
