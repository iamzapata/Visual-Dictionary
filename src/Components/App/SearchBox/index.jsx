import React, { Component } from "react"
import { func } from "prop-types"

class SearchBox extends Component {
  state = {
    inputValue: ""
  }

  onSubmitSearch = ev => {
    ev.preventDefault()
    const { inputValue } = this.state
    this.props.searchWord(inputValue)
  }

  onInputChange = ev => {
    const { value: inputValue } = ev.target
    this.setState({ inputValue })
  }

  render() {
    const { inputValue } = this.state
    return (
      <div className="SearchBox">
        <form
          className="SearchBox__Form"
          onSubmit={ev => this.onSubmitSearch(ev)}
        >
          <input
            type="text"
            className="SearchBox__Input"
            onChange={this.onInputChange}
            placeholder="Define ..."
            value={inputValue}
          />
          <button>Define</button>
        </form>
      </div>
    )
  }
}

SearchBox.propTypes = {
  searchWord: func.isRequired
}

export default SearchBox
