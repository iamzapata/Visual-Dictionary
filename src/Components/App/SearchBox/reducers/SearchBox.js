import ActionTypes from 'Components/App/SearchBox/actionTypes';

const defaultState = {
  isLoading: false,
  err: null,
  results: [],
};

export default function searchWord(state = defaultState, action) {
  const {
    type, isLoading, err, results,
  } = action;
  switch (type) {
    case ActionTypes.SEARCH_WORD_REQUEST:
      return {
        ...state,
        isLoading,
      };
    case ActionTypes.SEARCH_WORD_SUCCESS:
      return {
        ...state,
        isLoading,
        results,
      };
    case ActionTypes.SEARCH_WORD_FAILURE:
      return {
        ...state,
        isLoading,
        err,
      };
    default:
      return state;
  }
}
