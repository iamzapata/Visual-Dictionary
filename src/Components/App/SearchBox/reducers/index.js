import { combineReducers } from 'redux';
import SearchBox from './SearchBox';

const rootReducer = combineReducers({
  SearchBox,
});

export default rootReducer;
