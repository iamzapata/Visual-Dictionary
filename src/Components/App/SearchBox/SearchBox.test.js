import React from "react"
import {
  render,
  renderIntoDocument,
  cleanup,
  fireEvent
} from "react-testing-library"
import SearchBox from "./index.jsx"

describe("<SearchBox />", () => {
  afterEach(cleanup)

  it("Should render self", () => {
    const props = {
      searchWord: jest.fn()
    }
    const { container } = render(<SearchBox {...props} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("Should have a SearchBox className", () => {
    const props = {
      searchWord: jest.fn()
    }
    const { container } = render(<SearchBox {...props} />)
    const className = container.querySelectorAll(".SearchBox")
    expect(className.length).toEqual(1)
  })

  it("Should have a form element", () => {
    const props = {
      searchWord: jest.fn()
    }
    const { container } = render(<SearchBox {...props} />)
    const form = container.firstChild.querySelector("form")
    expect(form).toBeTruthy()
  })

  it("Should have an input element", () => {
    const props = {
      searchWord: jest.fn()
    }
    const { queryByPlaceholderText } = render(<SearchBox {...props} />)
    const input = queryByPlaceholderText("Define ...")
    expect(input).toBeTruthy()
  })

  it("Should have a submit button", () => {
    const props = {
      searchWord: jest.fn()
    }
    const { queryByText } = render(<SearchBox {...props} />)
    const submitButton = queryByText("Define")
    expect(submitButton).toBeTruthy()
  })

  it("Should allow the user to type a word to search for", () => {
    const searchString = "lion"
    const props = {
      searchWord: jest.fn()
    }
    const { queryByPlaceholderText } = renderIntoDocument(
      <SearchBox {...props} />
    )
    const inputNode = queryByPlaceholderText("Define ...")
    inputNode.value = "lion"
    fireEvent.change(inputNode)
    expect(queryByPlaceholderText("Define ...").value).toEqual(searchString)
  })

  it("Should allow the user to submit search by pressing enter key", () => {
    const searchWord = jest.fn()
    const props = {
      searchWord
    }
    const { queryByPlaceholderText } = renderIntoDocument(
      <SearchBox {...props} />
    )
    const input = queryByPlaceholderText("Define ...")
    fireEvent.submit(input)
    expect(searchWord).toHaveBeenCalledTimes(1)
  })

  it("Should allow the user to submit search by clicking define button", () => {
    const searchWord = jest.fn()
    const props = {
      searchWord
    }
    const { queryByText } = renderIntoDocument(<SearchBox {...props} />)
    const button = queryByText("Define")
    fireEvent.submit(button)
    expect(searchWord).toHaveBeenCalledTimes(1)
  })
})
