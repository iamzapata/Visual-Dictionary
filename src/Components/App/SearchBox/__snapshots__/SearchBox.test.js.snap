// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBox /> Should render self 1`] = `
<div
  class="SearchBox"
>
  <form
    class="SearchBox__Form"
  >
    <input
      class="SearchBox__Input"
      placeholder="Define ..."
      type="text"
      value=""
    />
    <button>
      Define
    </button>
  </form>
</div>
`;
